File: src/main/java/io/lsdconsulting/lsd/distributed/generator/diagram/data/InteractionDataGenerator.java
package io.lsdconsulting.lsd.distributed.generator.diagram.data;

import com.fasterxml.jackson.annotation.JsonInclude;
import io.lsdconsulting.lsd.distributed.access.model.InterceptedInteraction;
import lombok.Builder;
import lombok.Value;
import lsd.format.Parser;

import java.util.Collection;
import java.util.List;
import java.util.Map;

import static com.fasterxml.jackson.annotation.JsonInclude.Include.NON_EMPTY;
import static io.lsdconsulting.lsd.distributed.access.model.Type.*;

public class InteractionDataGenerator {

    public InteractionData buildFrom(InterceptedInteraction interceptedInteraction) {
        return InteractionData.builder()
                .requestHeaders(interceptedInteraction.getType().equals(REQUEST) ? interceptedInteraction.getRequestHeaders() : null)
                .responseHeaders(interceptedInteraction.getType().equals(RESPONSE) ? interceptedInteraction.getResponseHeaders() : null)
                .headers(List.of(PUBLISH, CONSUME).contains(interceptedInteraction.getType()) ? interceptedInteraction.getRequestHeaders() : null)
                .body(generateBody(interceptedInteraction.getBody()))
                .build();
    }

    private Object generateBody(String body) {
        Map<String, Object> bodyMap = Parser.parse(body);
        return bodyMap.isEmpty() ? body : bodyMap;
    }

    @Value
    @Builder
    static class InteractionData {
        @JsonInclude(NON_EMPTY)
        Map<String, Collection<String>> requestHeaders;

        @JsonInclude(NON_EMPTY)
        Map<String, Collection<String>> responseHeaders;

        @JsonInclude(NON_EMPTY)
        Map<String, Collection<String>> headers;

        Object body;
    }
}

File: src/main/kotlin/io/lsdconsulting/lsd/distributed/generator/diagram/data/InteractionDataGenerator.kt
package io.lsdconsulting.lsd.distributed.generator.diagram.data

import com.fasterxml.jackson.annotation.JsonInclude
import io.lsdconsulting.lsd.distributed.connector.model.InteractionType.*
import io.lsdconsulting.lsd.distributed.connector.model.InterceptedInteraction
import lsd.format.convertJsonStringToMap

fun buildDataFrom(interceptedInteraction: InterceptedInteraction) =
    InteractionData(
        requestHeaders = (if (interceptedInteraction.interactionType == REQUEST) interceptedInteraction.requestHeaders else null),
        responseHeaders = (if (interceptedInteraction.interactionType == RESPONSE) interceptedInteraction.responseHeaders else null),
        headers = (
                if (listOf(PUBLISH, CONSUME)
                        .contains(interceptedInteraction.interactionType)
                ) interceptedInteraction.requestHeaders else null
                ),
        body = (generateBody(interceptedInteraction.body))
    )

private fun generateBody(body: String?): Any? = convertJsonStringToMap(body).ifEmpty { body }

data class InteractionData(
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    var requestHeaders: Map<String, Collection<String>>?,

    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    var responseHeaders: Map<String, Collection<String>>?,

    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    var headers: Map<String, Collection<String>>?,
    var body: Any? = null
)
